forc-util/src/fs_locking.rs

```diff
@@ -0,0 +1,143 @@
+use crate::{hash_path, user_forc_directory};
+use std::{
+    fs::{create_dir_all, remove_file, File},
+    io::{self, Read, Write},
+    path::{Path, PathBuf},
+};
+
+/// Very simple AdvisoryPathMutex class
+///
+/// The goal of this struct is to signal other processes that a path is being used by another
+/// process exclusively.
+///
+/// This struct will self-healh if the process that locked the file is no longer running.
+pub struct PidFileLocking(PathBuf);
+
+impl PidFileLocking {
+    pub fn new<X: AsRef<Path>, Y: AsRef<Path>>(path: X, dir: Y, extension: &str) -> PidFileLocking {
```

Comment:
The path here is actually a filename right?

```suggestion
    pub fn new<X: AsRef<Path>, Y: AsRef<Path>>(filename: X, dir: Y, extension: &str) -> PidFileLocking {
```